src/sql/base/browser/ui/table/treeGrid.ts

```diff
@@ -0,0 +1,183 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import 'vs/css!./media/slick.grid';
+
+import { FilterableColumn, ITableConfiguration } from 'sql/base/browser/ui/table/interfaces';
+import { Table } from 'sql/base/browser/ui/table/table';
+import { IDisposableDataProvider } from 'sql/base/common/dataProvider';
+import { generateUuid } from 'vs/base/common/uuid';
+import { CellValueGetter, defaultCellValueGetter, defaultFilter, TableDataView } from 'sql/base/browser/ui/table/tableDataView';
+import { AsyncDataProvider } from 'sql/base/browser/ui/table/asyncDataView';
+import { StandardKeyboardEvent } from 'vs/base/browser/keyboardEvent';
+import { KeyCode } from 'vs/base/common/keyCodes';
+import { textFormatter, treeGridExpandableColumnFormatter } from 'sql/base/browser/ui/table/formatters';
+
+function defaultTreeGridFilter<T extends Slick.SlickData>(data: T[], columns: FilterableColumn<T>[], cellValueGetter: CellValueGetter = defaultCellValueGetter): T[] {
+	let filteredData = defaultFilter(data, columns, cellValueGetter);
+
+	// filtering out rows which have parent/grandparents collapsed.
+	filteredData = filteredData.filter((item) => {
+		let parent = data[item.parent];
+		while (parent) {
+			if (!parent.expanded) {
+				return false;
+			}
+			parent = data[parent.parent];
+		}
+		return true;
+	});
+
+	return filteredData;
+}
+
+/**
+ * TreeGrid component displays a hierarchical table data grouped into expandable and collapsible nodes.
+ */
+export class TreeGrid<T extends Slick.SlickData> extends Table<T> {
+	constructor(parent: HTMLElement, configuration?: ITableConfiguration<T>, options?: Slick.GridOptions<T>) {
+		super(parent, configuration, options);
+		this._tableContainer.setAttribute('role', 'treegrid');
+		if (this._columns[0]) {
+			this._columns[0].formatter = this._columns[0].formatter ? treeGridExpandableColumnFormatter(this._columns[0].formatter) : treeGridExpandableColumnFormatter(textFormatter);
```

Comment:
`this._columns[0].formatter = treeGridExpandableColumnFormatter(this._columns[0].formatter ?? textFormatter) ;`